### PR TITLE
R3 Final Review

### PR DESCRIPTION
**Pull request by ServiceDog**

### CODE REVIEW

**Comment by owlot** [comment]
## Documents
- [ ] ICD
  - [ ] Current interfaces (grpc & rest)
  - [ ] Current relevant document links
- [ ] SDD
  - [ ] Description of grpc & rest handler implementations
  - [ ] Current sequence diagrams where applicable
  - [ ] Current relevant document links
- [ ] CONOPS
  - [ ] Current information
  - [ ] Bibliography (if any citations)
- [ ] README (see svc-cargo for example)
  - [ ] Badges
  - [ ] Accurate installation instructions
  - [ ] Current links to other documents
  - [ ] Arrow DAO links
- [ ] Any instance of :construction: should genuinely be under construction

## Code (not applicable to examples/ folder)
- [ ] No unhandled unwrap()s
- [ ] Accurate (!) code comments on all functions
- [ ] No `# Arguments` and `# Returns` in function comments
- [ ] Function arguments checked where possible
- [ ] Sufficient logger (info, warn, and debug) messages
- [ ] Sensible Types (unsigned vs. signed)
- [ ] No `#[allow(dead_code)]` without good reason
- [ ] End all error, info, and debug messages with punctuation (periods, etc.)
- [ ] No FIXMEs or TODOs without a projected release (`TODO R4: etc.`)

## REST API
- [ ] Return codes accurately documented with openapi
- [ ] Return codes make sense
- [ ] Sufficiently descriptive error messages
- [ ] Do NOT reflect a user's bad input back at them in replies.
    - i.e. "Warning: '{}' is not a valid entry." should instead be "Warning: invalid entry."
    - Use debug messages to see the invalid argument.

## Cargo.toml(s)
- [ ] Package versions correct
- [ ] No minor versions specified
- [ ] Links to crates.io package, not git branch
- [ ] Must have `license-file = "LICENSE"` under `[package]`

File: docs/conops.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Concept of Operations - `svc-FIXME`
+# Concept of Operations - `svc-gis`
```
Comment (owlot) [inline docs/conops.md:1]: Start of document format. Should start with the banner (and should not be restricted to width).

From svc-template-rust:
```
![Arrow Banner](https://github.com/Arrow-air/tf-github/raw/main/src/templates/doc-banner-services.png)

# Concept of Operations - `svc-FIXME`

FIXME: Some of the headings below may not apply to your microservice. Remove if
so.

## :telescope: Overview

FIXME - What does this microservice do, in brief?

### Metadata

| Attribute     | Description                                                       |
| ------------- |-------------------------------------------------------------------|
| Maintainer(s) | [Services Team](https://github.com/orgs/Arrow-air/teams/services) |
| Stuckee       |  |
| Status        | Draft     

## :books: Related Documents

Document | Description
--- | ---
[High-Level Concept of Operations (CONOPS)](https://github.com/Arrow-air/se-services/blob/develop/docs/conops.md) | Overview of Arrow microservices.
[High-Level Interface Control Document (ICD)](https://github.com/Arrow-air/se-services/blob/develop/docs/icd.md)  | Interfaces and frameworks common to all Arrow microservices.
[Requirements - `svc-FIXME`](FIXME - generate a link specifically for this module's view in NocoDB) | Requirements and user stories for this microservice.
[Interface Control Document (ICD) - `svc-FIXME`](./icd.md) | Defines the inputs and outputs of this microservice.
[Software Design Document (SDD) - `svc-FIXME`](./sdd.md) | Specifies the internal activity of this microservice.
```

File: docs/icd.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Interface Control Document (ICD) - `svc-FIXME`
+# Interface Control Document (ICD) - `svc-gis`
```
Comment (owlot) [inline docs/icd.md:1]: Start of document format. Should start with the banner (and should not be restricted to width).

From svc-template-rust:
```
![Arrow Banner](https://github.com/Arrow-air/tf-github/raw/main/src/templates/doc-banner-services.png)

# Interface Control Document (ICD) - `svc-FIXME`

## :telescope: Overview

This document defines the gRPC and REST interfaces unique to the `svc-FIXME`
microservice.

### Metadata

| Attribute     | Description                                                       |
| ------------- |-------------------------------------------------------------------|
| Maintainer(s) | [Services Team](https://github.com/orgs/Arrow-air/teams/services) |
| Stuckee       |  |
| Status        | Draft                                                             |

## :books: Related Documents

Document | Description
--- | ---
[High-Level Concept of Operations (CONOPS)](https://github.com/Arrow-air/se-services/blob/develop/docs/conops.md) | Overview of Arrow microservices.
[High-Level Interface Control Document (ICD)](https://github.com/Arrow-air/se-services/blob/develop/docs/icd.md)  | Interfaces and frameworks common to all Arrow microservices.
[Requirements - `svc-FIXME`](FIXME - generate a link specifically for this module's view in NocoDB) | Requirements and user stories for this microservice.
[Concept of Operations - `svc-FIXME`](./conops.md) | Defines the motivation and duties of this microservice.
[Software Design Document (SDD) - `svc-FIXME`](./sdd.md) | Specifies the internal activity of this microservice.
```

File: docs/sdd.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Software Design Document (SDD) - `svc-FIXME` 
+# Software Design Document (SDD) - `svc-gis`
```
Comment (owlot) [inline docs/sdd.md:1]: Start of document format. Should start with the banner (and should not be restricted to width).

From svc-template-rust:
```
![Arrow Banner](https://github.com/Arrow-air/tf-github/raw/main/src/templates/doc-banner-services.png)

# Software Design Document (SDD) - `svc-FIXME`

## :telescope: Overview

This document details the software implementation of FIXME.

This service is responsible for FIXME

### Metadata

| Attribute     | Description                                                       |
| ------------- |-------------------------------------------------------------------|
| Maintainer(s) | [Services Team](https://github.com/orgs/Arrow-air/teams/services) |
| Stuckee       |  |
| Status        | Draft                                                             |

## :books: Related Documents

Document | Description
--- | ---
[High-Level Concept of Operations (CONOPS)](https://github.com/Arrow-air/se-services/blob/develop/docs/conops.md) | Overview of Arrow microservices.
[High-Level Interface Control Document (ICD)](https://github.com/Arrow-air/se-services/blob/develop/docs/icd.md)  | Interfaces and frameworks common to all Arrow microservices.
[Requirements - `svc-FIXME`](FIXME - generate a link specifically for this module's view in NocoDB) | Requirements and user stories for this microservice.
[Concept of Operations - `svc-FIXME`](./conops.md) | Defines the motivation and duties of this microservice.
[Interface Control Document (ICD) - `svc-FIXME`](./icd.md) | Defines the inputs and outputs of this microservice.
```

File: server/src/postgis/best_path.rs [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+                &request.node_uuid_end,
+                &request.time_start,
+                &request.time_end,
+            ],
+        )
+        .await
+    {
+        Ok(results) => Ok(results),
+        Err(e) => {
+            println!(
```
Comment (owlot) [inline server/src/postgis/best_path.rs:170]: Should be: `postgis_error!(`

File: server/src/postgis/best_path.rs [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+pub async fn best_path(
+    path_type: PathType,
+    request: BestPathRequest,
+    pool: deadpool_postgres::Pool,
+) -> Result<Vec<PathSegment>, PathError> {
+    let request = PathRequest::try_from(request)?;
+    let client = match pool.get().await {
+        Ok(client) => client,
+        Err(e) => {
+            println!("(best_path) could not get client from pool.");
```
Comment (owlot) [inline server/src/postgis/best_path.rs:198]: Should be: `postgis_error!(`

File: server/src/postgis/nearest.rs [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                &start_node_id,
+                &request.limit,
+                &(request.max_range_meters as f64),
+            ],
+        )
+        .await
+    {
+        Ok(results) => Ok(results),
+        Err(e) => {
+            println!(
```
Comment (owlot) [inline server/src/postgis/nearest.rs:95]: Should be: `postgis_error!(`

File: server/src/postgis/nearest.rs [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+                &request.start_node_id,
+                &request.limit,
+                &(request.max_range_meters as f64),
+            ],
+        )
+        .await
+    {
+        Ok(results) => Ok(results),
+        Err(e) => {
+            println!(
```
Comment (owlot) [inline server/src/postgis/nearest.rs:124]: Should be: `postgis_error!(`

File: docs/sdd.md
Comment (owlot) [inline docs/sdd.md:1]: Missing "💬 gRPC Handlers" block with sequence diagrams.
Please add it with maybe a "🚧 TODO(R4)" comment if it's planned to be added later.

**Review by owlot — COMMENTED** [review]
`client-grpc/README.md` is missing content (header etc).

From svc-template-rust:
```
![Arrow Banner](https://github.com/Arrow-air/tf-github/raw/main/src/templates/doc-banner-services.png)

# svc-template-rust gRPC Client

![GitHub stable release (latest by date)](https://img.shields.io/github/v/release/Arrow-air/svc-template-rust?sort=semver&color=green) ![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/Arrow-air/svc-template-rust?include_prereleases) [![Coverage Status](https://coveralls.io/repos/github/Arrow-air/svc-template-rust/badge.svg?branch=develop)](https://coveralls.io/github/Arrow-air/svc-template-rust)
![Sanity Checks](https://github.com/arrow-air/svc-template-rust/actions/workflows/sanity_checks.yml/badge.svg?branch=develop) ![Python PEP8](https://github.com/arrow-air/svc-template-rust/actions/workflows/python_ci.yml/badge.svg?branch=develop) ![Rust Checks](https://github.com/arrow-air/svc-template-rust/actions/workflows/rust_ci.yml/badge.svg?branch=develop) 
![Arrow DAO Discord](https://img.shields.io/discord/853833144037277726?style=plastic)

## Overview

Exposes svc-template-rust gRPC client functions
```

File: server/README.md [hunk 8]
```diff
@@ -1,22 +1,15 @@
...
+![Sanity Checks](https://github.com/arrow-air/svc-gis/actions/workflows/sanity_checks.yml/badge.svg?branch=main)
+![Rust Checks](https://github.com/arrow-air/svc-gis/actions/workflows/rust_ci.yml/badge.svg?branch=main)
+![Python PEP8](https://github.com/arrow-air/svc-gis/actions/workflows/python_ci.yml/badge.svg?branch=main)
 ![Arrow DAO
 Discord](https://img.shields.io/discord/853833144037277726?style=plastic)
 
-2. *Grep for 'template' in all files, replace with name of this service*
-3. *Remove this and previous numbered bullets*
-
 ## :telescope: Overview
```
[the remark this answers is not in the document]
Comment (ServiceDog) [reply server/README.md]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: docs/conops.md (hunk 1 above)
Comment (ServiceDog) [reply docs/conops.md]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: docs/icd.md (hunk 2 above)
Comment (ServiceDog) [reply docs/icd.md]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: docs/sdd.md (hunk 3 above)
Comment (ServiceDog) [reply docs/sdd.md]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: server/src/postgis/best_path.rs [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+                &request.node_uuid_end,
+                &request.time_start,
+                &request.time_end,
+            ],
+        )
+        .await
+    {
+        Ok(results) => Ok(results),
+        Err(e) => {
+            println!(
```
[the remark this answers is not in the document]
Comment (ServiceDog) [reply server/src/postgis/best_path.rs]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: server/src/postgis/best_path.rs (hunk 4 above)
Comment (ServiceDog) [reply server/src/postgis/best_path.rs]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: server/src/postgis/best_path.rs (hunk 5 above)
Comment (ServiceDog) [reply server/src/postgis/best_path.rs]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: server/src/postgis/nearest.rs (hunk 6 above)
Comment (ServiceDog) [reply server/src/postgis/nearest.rs]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: server/src/postgis/nearest.rs (hunk 7 above)
Comment (ServiceDog) [reply server/src/postgis/nearest.rs]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

File: docs/sdd.md
Comment (ServiceDog) [reply docs/sdd.md:1]: https://github.com/Arrow-air/svc-gis/commit/2b93948539015c3003bde9fba7385a81e9e6f4dd

**Review by owlot — APPROVED** [review]

**Comment by owlot** [comment]
Awesome work on this service!

**Pull request merged**